### DISCUSSION TITLE
Is it possible to use a preview F# version?

### DISCUSSION THREAD

**Discussion by bmitc** [dotnet/interactive#3100] [Q&A]
I am trying to take advantage of an F# feature that is in the preview version. The feature is setting static members on interfaces. I was told that to get this preview version on the command line, I could do `dotnet fsi --langversion:preview`. Is there a way to do this with Polyglot Notebooks?

I am trying to define an interface that requires this sort of declaration, and I'd love to do it in a notebook rather than in a project.

Note: I am not entirely clear what is meant by "preview" version of F#. I have the latest .NET 8 preview installed, but I can only get this feature in FSI if I run the above command.

I searched and see this https://github.com/dotnet/interactive/issues/1408, so it appears that maybe it's not possible, but I wanted to make sure.

Thank you!

**Comment by jonsequitur** [comment]
If this is possible at the FSI level then it should be fairly simple to support in Polyglot Notebooks. @vzarytovskii any thoughts on that?

**Reply by vzarytovskii** [reply]
Yeah, I would think that it'll be fairly easy.


 @KevinRansom will probably know better.

**Comment by KevinRansom** [comment]
F# in notebooks should already be preview mode of F#:

https://github.com/dotnet/interactive/blob/d0fb7fd58c1d3575bb7697e7a7c3c133a70ff59d/src/Microsoft.DotNet.Interactive.FSharp/FSharpKernel.fs#L46

The version of F# included in notebooks main branch is :  43.7.200  which is 5 months old.
https://github.com/dotnet/interactive/blob/d0fb7fd58c1d3575bb7697e7a7c3c133a70ff59d/Directory.Build.targets#L38

The latest release is somewhat more recent:
https://www.nuget.org/packages/FSharp.Compiler.Service/43.8.100-preview.23327.15

Notebooks should probably update to 43.7.300 which is also in the latest VS RTM release ... and the VSCode preview version would ideally be on [43.8.100-preview.23327.15](https://www.nuget.org/packages/FSharp.Compiler.Service/43.8.100-preview.23327.15)

**Reply by bmitc** [reply]
I don't know why I didn't include this in my question, but I am specifically trying to do:

```fsharp
[<Interface>]
type ITesting =
    static member Testing x = x
```

and this currently doesn't work, even for a preview version of the VS Code Polyglot Notebooks extension. It actually doesn't give the same warning that .NET 7 and the .NET 8 Preview do (which say to use "language version 'PREVIEW'". This is the error:

<img width="644" alt="image" src="https://github.com/dotnet/interactive/assets/65685447/fe27a875-6c00-463a-abf2-6f5b5d18956e">

The static member in interfaces feature was discussed in https://github.com/fsharp/fslang-suggestions/issues/1191, and I asked how to use the feature at the bottom of that PR.

**Reply by bmitc** [reply]
Does this likely mean that this feature was added between the last F# version included in the notebooks branch and the current one? .NET 7 gives the "PREVIEW" error message as well, so I'm guessing it was patched with the error message knowing this feature is coming in .NET 8?

**Reply by vzarytovskii** [reply]
It should include the preview flag https://github.com/dotnet/interactive/blob/d0fb7fd58c1d3575bb7697e7a7c3c133a70ff59d/src/Microsoft.DotNet.Interactive.FSharp/FSharpKernel.fs#L46, so not sure why it doesn't light up preview features.

Cc @jonsequitur @colombod

**Reply by jonsequitur** [reply]
We're not using a preview version of `FSharp.Compiler.Service`.

**Reply by bmitc** [reply]
So it's sounding like it's not possible then at the moment? I should just wait?

In general, I don't even know what "language version 'PREVIEW'" means in the F# error message, even now, other than I can tell FSI about it via a command line option. I don't understand how it relates to the preview versions of .NET (I thought F# was synced to .NET releases now). It sounds like it's more of a "mode" than a version? Because I'm not sure what other version of F# is being used in .NET 8 Preview other than a preview version already. But these questions aren't necessarily needed to be answered here, as my main concern right now is using this "preview" version or mode in Polyglot Notebooks, if possible. (But if anyone knows the answer to the "preview" vs version thing, it would be helpful!)
